compiler/src/emit.rs

```diff
@@ -0,0 +1,101 @@
+use analyzer::relations::Symbol;
+use analyzer::types::hir::{Declaration, ExprKind, TypedExpr};
+use analyzer::types::*;
+use ast::value::LiteralValue;
+
+use crate::bytecode::{Bytecode, Opcode};
+use crate::constant_pool::ConstantPool;
+use crate::emit::invoke::emit_process_call;
+use crate::emit::jump::{emit_break, emit_conditional, emit_continue, emit_loop};
+
+mod invoke;
+mod jump;
+
+pub struct EmissionState {
+    // the start instruction position of the enclosing loop
+    // set to 0 if there is no loop
+    pub enclosing_loop_start: usize,
+    // All the placeholders waiting for the end of the loop.
+    // When the loop compilation ends, all those placeholder are filled with the
+    // first instruction pointer after the loop.
+    pub enclosing_loop_end_placeholders: Vec<usize>,
+}
+
+impl EmissionState {
+    pub fn new() -> Self {
+        Self {
+            enclosing_loop_start: 0,
+            enclosing_loop_end_placeholders: Vec::new(),
+        }
+    }
+}
+
+fn emit_literal(literal: &LiteralValue, emitter: &mut Bytecode, cp: &mut ConstantPool) {
+    match literal {
+        LiteralValue::String(string) => {
+            let str_ref = cp.insert_string(string.clone());
+            emitter.emit_string_constant_ref(str_ref);
+        }
+        LiteralValue::Int(integer) => {
+            emitter.emit_int(*integer);
+        }
+        LiteralValue::Float(f) => {
+            emitter.emit_float(*f);
+        }
+    }
+}
+
+fn emit_ref(symbol: &Symbol, emitter: &mut Bytecode) {
+    match symbol {
+        Symbol::Local(id) => emitter.emit_get_local(id.0 as u8),
+        _ => todo!(),
+    }
+}
+
+fn emit_declaration(
+    declaration: &Declaration,
+    emitter: &mut Bytecode,
+    cp: &mut ConstantPool,
+    state: &mut EmissionState,
+) {
+    let value = declaration
+        .value
+        .as_ref()
+        .expect("var declaration without initializer not supported");
+    emit(value, emitter, cp, state);
+    emitter.emit_code(Opcode::SetLocal);
+    emitter.bytes.push(declaration.identifier.0 as u8);
```

Comment:
```suggestion
    if let Some(value) = &declaration.value {
        emit(value, emitter, cp, state);
        emitter.emit_code(Opcode::SetLocal);
        emitter.bytes.push(declaration.identifier.0 as u8);
    }
```
This is a regression.
https://github.com/moshell-lang/moshell/blob/8823ae3e8236d1777fde0e2d7d2a5e48021e00da/compiler/src/lib.rs#L15
On the master branch, the initializer is pattern matched so that it supports variables without initializers. This is only a concern for the analyzer.